src/cargo/ops/registry/publish.rs

```diff
@@ -57,151 +64,192 @@ pub struct PublishOpts<'gctx> {
-        }
-        val => val,
-    };
     let source_ids = super::get_source_id(opts.gctx, reg_or_index.as_ref())?;
     let mut registry = super::registry(
         opts.gctx,
         &source_ids,
         opts.token.as_ref().map(Secret::as_deref),
         reg_or_index.as_ref(),
         true,
-        Some(operation).filter(|_| !opts.dry_run),
+        Some(Operation::Read).filter(|_| !opts.dry_run),
     )?;
-    verify_dependencies(pkg, &registry, source_ids.original)?;
 
-    // Prepare a tarball, with a non-suppressible warning if metadata
-    // is missing since this is being put online.
-    let tarball = ops::package_one(
+    // Validate all the packages before publishing any of them.
+    for (pkg, _) in &pkgs {
+        verify_dependencies(pkg, &registry, source_ids.original)?;
+    }
+
+    let specs = Packages::from_flags(
+        false,
+        vec![],
+        just_pkgs.iter().map(|p| p.name().to_string()).collect(),
+    )?;
+    let pkg_dep_graph = ops::cargo_package::package_with_dep_graph(
         ws,
-        pkg,
         &PackageOpts {
             gctx: opts.gctx,
             verify: opts.verify,
             list: false,
             check_metadata: true,
             allow_dirty: opts.allow_dirty,
-            to_package: Packages::Default,
+            to_package: specs,
             targets: opts.targets.clone(),
             jobs: opts.jobs.clone(),
             keep_going: opts.keep_going,
-            cli_features,
-            reg_or_index,
+            cli_features: opts.cli_features.clone(),
+            reg_or_index: reg_or_index.clone(),
         },
     )?;
 
-    if !opts.dry_run {
-        let hash = cargo_util::Sha256::new()
-            .update_file(tarball.file())?
-            .finish_hex();
-        let operation = Operation::Publish {
-            name: pkg.name().as_str(),
-            vers: &ver,
-            cksum: &hash,
-        };
-        registry.set_token(Some(auth::auth_token(
-            &opts.gctx,
-            &source_ids.original,
-            None,
-            operation,
-            vec![],
-            false,
-        )?));
-    }
+    let (mut order, mut ready) = PublishOrder::new(&pkg_dep_graph.graph);
+    let mut to_confirm: HashSet<_> = ready.iter().copied().collect();
+    let mut remaining: HashSet<_> = just_pkgs.iter().map(|pkg| pkg.package_id()).collect();
 
-    opts.gctx
-        .shell()
-        .status("Uploading", pkg.package_id().to_string())?;
-    transmit(
-        opts.gctx,
-        ws,
-        pkg,
-        tarball.file(),
-        &mut registry,
-        source_ids.original,
-        opts.dry_run,
-    )?;
-    if !opts.dry_run {
-        const DEFAULT_TIMEOUT: u64 = 60;
-        let timeout = if opts.gctx.cli_unstable().publish_timeout {
-            let timeout: Option<u64> = opts.gctx.get("publish.timeout")?;
-            timeout.unwrap_or(DEFAULT_TIMEOUT)
+    while !to_confirm.is_empty() {
+        if !ready.is_empty() {
```

Comment:
Why do we continue the loop when `ready` is empty?

Reply:
Say we have packages `main`, `dep1`, and `dep2`, where `main` depends on both of the `dep`s. The first time through the loop, both of the deps are ready, so we upload them both. But now suppose that when we poll, we only see `dep1` as ready. Then we'll run through the loop again even though nothing is ready for upload.

Having said that, I'm going to remove this check in favor of printing the "Uploading..." messages one-by-one for more precise context.

Reply:
> Say we have packages main, dep1, and dep2, where main depends on both of the deps. The first time through the loop, both of the deps are ready, so we upload them both. But now suppose that when we poll, we only see dep1 as ready. Then we'll run through the loop again even though nothing is ready for upload.

That we are doubling the timeout for some packages is non-obvious and this should probably have a comment in the code.

Reply:
The comment that was added to resolve this, while good and covering literally what I asked for, didn't fully address the concern I had. The comment was put deeper, inside of `if 0 < timeout {` but this doesn't help someone to understand why you can have `remaining` without `ready` where they are reading this non-obvious code.

Reply:
Ah, since I'd removed the `!ready.is_empty()` check it wasn't clear to me that the explanation was still needed. I've added one, at the point where we retrieve the ready packages.